proposals/0652-import-shadowing.rst

```diff
@@ -258,43 +258,70 @@ something like
 
  module A (foo) where
 
- import M -- This exports "foo"
+ import M (foo)
 
  foo = ...
 
 then the ``foo`` exported by ``A`` should be the one defined in
 ``A``'s top-level.
 
-When modules are reexported wholesale, shadowing doesn't come into
-play, and so we keep the behaviour without this extension: the form
-``module M`` names the set of all entities that are in scope with both
-an unqualified name ``e`` and a qualified name ``M.e``. Example:
+.. _module-reexports:
+
+Module reexports
+^^^^^^^^^^^^^^^^
+
+In an export list, `Paragraph 5 of Section 5.2 of the Haskell 2010 report
+<https://www.haskell.org/onlinereport/haskell2010/haskellch5.html#x11-1000005.2>`_
+specifies that the form ``module M`` names the set of all entities
+that are in scope with both an unqualified name ``e`` and a qualified
+name ``M.e``.  So in this example, *without* ``ImportShadowing``:
 
 ::
 
  module A (module M) where
 
- import M -- this exports "foo"
+ import M (foo, wombat)
+ import N (foo)
 
- foo = ...
+ foo = True
+
+the ``module M`` exports ``M.foo`` because that entity is in scope
+both as ``M.foo`` and with unqualified name ``foo``.  The fact that
+``N.foo`` and ``A.foo`` are *also* in scope with unqualified name
+``foo`` does not matter.
 
-Here, it is ``M.foo`` that is (re-)exported by ``A``, not ``A.foo``.
+With ``ImportShadowing``, however, the local definition of ``A.foo``
+*completely hides* the unqualified imports of ``M.foo`` and ``N.foo``;
+so now ``M.foo`` is no longer in scope with unqualified name ``foo``;
+so the ``module M`` export list exports only ``M.wombat``.
```

Comment:
If I'm reading correctly, this is a backward incompatible effect of the extension (albeit a bit of a corner case). It should be documented in the backward compatibility section.

Reply:
But it's only if the extension is enabled.

Reply:
Indeed. But we've been considering it part of the game to document whether enabling the extension rejects (or change the semantics) of existing programs or not as part of the backward-compatibility section.

Quite frankly, I'm anticipating to want this extension enabled in the next GHC edition. And such explicit documentation will help with documenting how to transition to said next GHC edition.

Reply:
Arnaud is right.  In back-compat we identify cases where
* An *existing* program will break
*  when the extension is enabled

Let's just document this point.  It's a case that is vanishingly unlikely to arise in practice.